### ISSUE TITLE
Add "xtol" and "ftol

### ISSUE THREAD

**Issue by MichaelClerx** [pints-team/pints#1502]

**Comment by mirams** [comment]
I'm happy with those proposals, doing absolute ones makes more sense as relative would presumably be with respect to initial guess or something which would be a bit strange to vary run to run. Haven't ever seen a package specify a vector of xtol, but it makes sense, maybe I have just been automatically doing `[tol]*n_params` behind the scenes, but a nice friendly message saying "the xtol should be a vector of length n_parameters specifying separate tolerances for each parameter" would make it easy enough to see what needs to be done.

I think matlab's `optimset` (common optimisation options object used by a range of optimisers) is the thing to look at, that uses 'TolX' and 'TolFun' nowadays: [optimset docs](https://uk.mathworks.com/help/matlab/ref/optimset.html?searchHighlight=optimset&s_tid=srchtitle_support_results_1_optimset)

**Comment by MichaelClerx** [comment]
OK, so with our `threshold` then we can match all 3. That's good!

**Comment by MichaelClerx** [comment]
@mirams @martinjrobins @chonlei discussion point: The `ftol` is essentially the same as `set_max_unchanged_iterations` with `iterations=1`.

Options:

1. Don't implement a seperate `ftol` criterion
2. Implement a separate ftol criterion
3. Implement a method `set_function_tolerance` that internally calls `set_max_unchanged_iterations` (and document that this is what it does)

Input from anyone else doing optimisation appreciated!

**Comment by MichaelClerx** [comment]
There is a case for having both I think, if you want to use the `ftol` as a convergence test for something that is expected to be asymptotic at the end but keep that max-unchanged to detect being stuck

**Comment by chonlei** [comment]
I am not too sure. For option 3, it is like having an alias to `set_max_unchanged_iterations` with argument `iterations=1`, right? If it is what it is, then might be the easiest workaround?

**Comment by mirams** [comment]
Yeah, an argument for just wrapping, so `ftol(tol) = set_max_unchanged_iterations(iterations=1, tol)`. I guess the problem there is that it somewhat needs all the optimisers to agree what an 'iteration' means. I had imagined that you'd want to be usually wrapping the `ftol()` function provided by off-the-shelf optimisers, and I imagine different optimisers can and will define "over how many function evals/iterations/whatever" depending on how they work and are known to converge. Whereas I guess the `max_changed_iterations` is something we're implementing ourselves?

I think confusion could then arise if the user can call both, but perhaps not, might even be a good idea to be able to set a collection of these like
```
set_max_unchanged_iterations(iterations=1, tol=1e-12)
set_max_unchanged_iterations(iterations=10, tol=1e-10)
set_max_unchanged_iterations(iterations=100, tol=1e-8)
```
and stop the first time one is met!!

**Comment by MichaelClerx** [comment]
> I guess the problem there is that it somewhat needs all the optimisers to agree what an 'iteration' means. I had imagined that you'd want to be usually wrapping the ftol() function provided by off-the-shelf optimisers, and I imagine different optimisers can and will define "over how many function evals/iterations/whatever" depending on how they work and are known to converge

No we ignore what the optimisers say so that we can use the same controller to run optimisations with them! They use different number of evaluations per iteration, that's fine :D

**Comment by MichaelClerx** [comment]
> might even be a good idea to be able to set a collection of these like

That's an interesting idea! Would require some laborious unsetting syntax though, e.g. `unset_unchanged_iterations` with the exact same arguments, or having `set` return some object you can then pass in to unset.

Will leave that for now as we don't have a direct use case :D

**Comment by MichaelClerx** [comment]
But both leaning towards option 2 then?
